src/mlpack/tests/ham_test.cpp

```diff
@@ -0,0 +1,280 @@
+  TreeMemory<double> mem(1, 1, J, W);
+  std::vector<std::vector<double>> initMemSTL = {{0}};
+  arma::mat initMem = convertToArma(initMemSTL);
+  mem.Initialize(initMem);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, 1);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).at(0, 0), 0.);
+  mem.Update(0, arma::mat("12."));
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, 1);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).at(0, 0), 12.);
+}
+
+BOOST_AUTO_TEST_CASE(TreeMemoryTestMinimumNDim)
+{
+  /*AddJoiner<double> J;
+  ReplaceWriter<double> W;*/
+  size_t memSize = 5;
+  FFN<MeanSquaredError<> > J = AddJoiner(memSize), W = ReplaceWriter(memSize);
+  TreeMemory<double> mem(
+      1, memSize, J, W);
+  std::vector<std::vector<double>> initMemSTL = {{0, 0, 0, 0, 0}};
+  arma::mat initMem = convertToArma(initMemSTL);
+  mem.Initialize(initMem);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, memSize);
+  for (size_t i = 0; i < memSize; ++i) {
+    BOOST_REQUIRE_EQUAL(mem.Leaf(0).at(i, 0), 0.);
+  }
+  mem.Update(0, 12 * arma::ones(memSize, 1));
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, memSize);
+  for (size_t i = 0; i < memSize; ++i)
+    BOOST_REQUIRE_EQUAL(mem.Leaf(0).at(i, 0), 12.);
+}
+
+BOOST_AUTO_TEST_CASE(TreeMemoryTestPowerOfTwo)
+{
+  /*AddJoiner<double> J;
+  ReplaceWriter<double> W;*/
+  FFN<MeanSquaredError<> > J = AddJoiner(1), W = ReplaceWriter(1);
+  TreeMemory<double> mem(8, 1, J, W);
+  std::vector<std::vector<double>> initMemSTL =
+      {{0}, {0}, {0}, {0}, {0}, {0}, {0}, {0}};
+  arma::mat initMem = convertToArma(initMemSTL);
+  mem.Initialize(initMem);
+  for (size_t idx = 0; idx < 15; ++idx)
+    BOOST_REQUIRE_EQUAL(mem.Cell(idx).at(0, 0), 0);
+  mem.Update(0, arma::mat("1."));
+  mem.Update(1, arma::mat("2."));
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, 1);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).at(0, 0), 1);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, 1);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(1).at(0, 0), 2);
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.LeafIndex(0))).at(0, 0), 3);
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.Parent(mem.LeafIndex(0)))).at(0, 0), 3);  
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.Parent(mem.Parent(mem.LeafIndex(0))))).at(0, 0), 3); 
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.LeafIndex(0))).n_elem, 1);
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.Parent(mem.LeafIndex(0)))).n_elem, 1);  
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.Parent(mem.Parent(mem.LeafIndex(0))))).n_elem, 1); 
+}
+
+BOOST_AUTO_TEST_CASE(TreeMemoryTestPowerOfTwoNDim)
+{
+  /*AddJoiner<double> J;
+  ReplaceWriter<double> W;*/
+  size_t memSize = 4;
+  FFN<MeanSquaredError<> > J = AddJoiner(memSize), W = ReplaceWriter(memSize);
+  TreeMemory<double> mem(
+      8, memSize, J, W);
+  std::vector<std::vector<double>> initMemSTL =
+      {{0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0},
+       {0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}};
+  arma::mat initMem = convertToArma(initMemSTL);
```

Comment:
Couldn't you just do, e.g.,

```
arma::mat matrix = {{1, 2, 3, 4}, {5, 6, 7, 8}};
matrix = matrix.t();
```

I think that gives you the same result, and uses significantly less code (and `convertToArma()` can be removed).

Also, in the particular case above this comment here, I think you can just do

```
arma::mat initMem(4, 8, arma::fill::zeros);
```

or if you really want to have it be row-major when you initialize it,

```
arma::mat initMem(8, 4, arma::fill::zeros);
initMem = initMem.t();
```

I'd advise treating Armadillo matrices in C++ as column-major, even if you find it a little counterintuitive, because the rest of the library does.  One workaround would be to add a comment clarifying what rows represent and what columns represent.  i.e. with `arma::mat m(8,
 4, arma::fill::zeros)` you could write a comment like `// 4 points in 8 dimensions.` or similar.  (For what it's worth: every textbook in the world implies row-major, but FORTRAN and the low-level scientific computing world imply column-major.  I think even numpy stores matrices in column-major format like Armadillo by default.)